src/sampling.rs

```diff
@@ -85,147 +79,13 @@ where
-    rng: R,
-}
-
-impl<R> WeightedRangeGenerator<R>
-where
-    R: Rng,
-{
-    #[allow(dead_code)]
-    pub fn new(rng: R, weights: &[usize]) -> WeightedRangeGenerator<R> {
-        assert!(!weights.is_empty(), "Cannot sample from zero elements.");
-
-        let mut prefix_sum = Vec::with_capacity(weights.len());
-        let mut sum = 0;
-        for &v in weights {
-            sum += v;
-            prefix_sum.push(sum);
-        }
-
-        WeightedRangeGenerator {
-            prefix_sum,
-            upper: sum + 1,
-            rng,
-        }
-    }
-}
-
-impl<R> Iterator for WeightedRangeGenerator<R>
-where
-    R: Rng,
-{
-    type Item = usize;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let val = self.rng.gen_range(1..self.upper);
-
-        let idx = match self.prefix_sum.binary_search(&val) {
-            Ok(idx) => idx,
-            Err(idx) => idx,
-        };
-
-        Some(idx)
-    }
-}
-
-impl<R> RangeGenerator for WeightedRangeGenerator<R>
-where
-    R: Rng,
-{
-    fn upper_bound(&self) -> usize {
-        self.prefix_sum.len()
-    }
-}
-
-/// An iterator that draws from *[0, n)* with a Zipfian distribution.
-///
-/// This iterator returns integers from *[0, n)*, where the probability of
-/// each integer follows a Zipfian distribution.
-///
-/// This generator can be used to draw words from a vocabulary sorted by
-/// descending frequency. Since the token frequencies (presumably) have a
-/// Zipfian distribution, this will pick a token with a probability that
-/// is proportional to its frequency.
-#[allow(dead_code)]
-pub struct ZipfRangeGenerator<R> {
-    upper_bound: usize,
-    exponent: f64,
-    rng: R,
-    dist: ZipfDistribution,
-}
-
-impl<R> Clone for ZipfRangeGenerator<R>
-where
-    R: Clone,
-{
-    fn clone(&self) -> Self {
-        ZipfRangeGenerator {
-            upper_bound: self.upper_bound,
-            exponent: self.exponent,
-            rng: self.rng.clone(),
-            dist: ZipfDistribution::new(self.upper_bound, self.exponent).unwrap(),
-        }
-    }
-}
-
-impl<R> ZipfRangeGenerator<R>
-where
-    R: Rng,
-{
-    #[allow(dead_code)]
-    pub fn new(rng: R, upper: usize) -> Self {
-        Self::new_with_exponent(rng, upper, ZIPF_RANGE_GENERATOR_DEFAULT_EXPONENT)
-    }
-
-    pub fn new_with_exponent(rng: R, upper_bound: usize, exponent: f64) -> Self {
-        ZipfRangeGenerator {
-            upper_bound,
-            exponent,
-            rng,
-            dist: ZipfDistribution::new(upper_bound, exponent).unwrap(),
-        }
-    }
-}
-
-impl<R> Iterator for ZipfRangeGenerator<R>
-where
-    R: Rng,
-{
-    type Item = usize;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let r = self.dist.sample(&mut self.rng);
-        Some(r - 1)
-    }
-}
-
-impl<R> RangeGenerator for ZipfRangeGenerator<R>
-where
-    R: Rng,
-{
-    fn upper_bound(&self) -> usize {
-        self.upper_bound
+        self.vocab_indices.last().map(|l| l + 1).unwrap_or(0)
```

Comment:
the indices are sorted most frequent to least frequent, so the last value in `vocab_indices` is also the index of the least frequent word in the vocabulary?

Reply:
Indeed! Though since the array size is preconfigured, it could theoretically happen that the least frequent word(s) do not occur in the array. But I think `upper_bound` is still correct in that case, it should be the upper bound of what is generated by the range generator.

Though this method is only used by the unit tests, so the trait should probably not be `pub`.